lib/ansible/module_utils/vmware.py

```diff
@@ -32,6 +35,117 @@
     HAS_PYVMOMI = False
 
 
+class PyVmomiHelper(object):
+
+    def __init__(self, module):
+        if not HAS_PYVMOMI:
+            module.fail_json(msg='pyvmomi module required')
+
+        self.module = module
+        self.params = module.params
+        self.content = connect_to_api(self.module)
+
+    @staticmethod
+    def construct_vm_searchpath(datacenter, folder):
+        # https://pubs.vmware.com/vi3/sdk/ReferenceGuide/vim.SearchIndex.html#findByInventoryPath
+        #   "My Folder/My Datacenter/vm/Discovered VM/VM1"
+        #   * leading slash or trailing slash is ignored
+        #   * the following paths all represents the same object:
+        #       "a/b", "/a/b", "a/b/", and '/a/b/'
+        #   * Slashes in names must be represented using %2f
+
+        if not folder:
+            folder = '/vm'
+        else:
+            if not folder.startswith('/vm') and not folder.startswith('vm'):
+                folder = os.path.join('/vm', folder)
+
+        if folder.endswith('/'):
+            folder = folder.rstrip('/')
+
+        searchpath = os.path.join(datacenter, folder)
+        return searchpath
+
+    def getvm(self, name=None, uuid=None, searchpath=None):
+        '''Find a VM by name and optionally a folder with absolute certainty'''
+
+        vm = None
+
+        # prefer a UUID
+        if uuid is not None:
+            vm = find_vm_by_id(self.content, uuid, vm_id_type="uuid", instanceUuid=False)
+            return vm
+
+        # use a searchpath + FindByInventoryPath to search by whatever
+        # folder the user thinks it is in.
+        elif searchpath:
+
```

Comment:
Do you need to get all of the objects and inspect them or could you go straight to the VM?

```python
elif searchpath and name:
    vm_path = searchpath + '/' + name
    vm = self.content.searchIndex.FindByInventoryPath(vm_path)
```